File: go/non_modifying_sequence.go
package main

/**
 * Go doesn't support generic. We can fake it with reflectiion or type deduction.
 * But these methods will make code dirty and duplicate, so I just write a version
 * for int
 */
func AllOf(vi []int, f func(int) bool) bool {
	for _, v := range vi {
		if !f(v) {
			return false
		}
	}
	return true
}

func AnyOf(vi []int, f func(int) bool) bool {
	for _, v := range vi {
		if f(v) {
			return true
		}
	}
	return false
}

func NoneOf(vi []int, f func(int) bool) bool {
	for _, v := range vi {
		if f(v) {
			return false
		}
	}
	return true
}

func ForEach(vi []int, f func(int)) {
	for _, v := range vi {
		f(v)
	}
}

func Count(vi []int) int {
	return len(vi)
}

func CountIf(vi []int, f func(int) bool) int {
	counter := 0
	for _, v := range vi {
		if f(v) {
			counter++
		}
	}
	return counter
}

func Mismatch(v1 []int, v2 []int, f2 func(int, int) bool) (mismatchInV1 int,
	mismatchInV2 int, hasMisMatch bool) {
	BoundOfV2 := len(v2)
	for i, v := range v1 {
		if i > BoundOfV2 {
			break
		}
		if !f2(v, v2[i]) {
			return v, v2[i], true
		}
	}
	return 0, 0, false
}

func Equal(v1 []int, v2 []int, f2 func(int, int) bool) bool {
	if len(v1) != len(v2) {
		return false
	} else {
		for i, v := range v1 {
			if !f2(v, v2[i]) {
				return false
			}
		}
	}
	return true
}

func Find(vi []int, t int) int {
	for i, v := range vi {
		if v == t {
			return i
		}
	}
	return -1
}

func FindIf(vi []int, f func(int) bool) int {
	for i, v := range vi {
		if f(v) {
			return i
		}
	}
	return -1
}

func FindIfNot(vi []int, f func(int) bool) int {
	for i, v := range vi {
		if !f(v) {
			return i
		}
	}
	return -1
}

// find first element in sub-scope occurs in given scope
func FindFirstOf(scope []int, subScope []int,
	f2 func(int, int) bool) (index int, found bool) {

	for i := range scope {
		for _, v := range subScope {
			if f2(scope[i], v) {
				return i, true
			}
		}
	}
	return -1, false
}

func FindEnd(scope []int, subScope []int,
	f2 func(int, int) bool) (index int, found bool) {

	lenOfScope := len(scope)
	lenOfSubScope := len(subScope)
	if lenOfScope < lenOfSubScope {
		return -1, false
	}
	i := lenOfScope - lenOfSubScope
	for {
		for j := 0; ; j++ {
			if j == lenOfSubScope {
				return i, true
			}
			if !f2(scope[i+j], subScope[j]) {
				if i > 0 {
					i--
					break
				} else {
					return -1, false
				}
			}
		}

	}
}

func AdjacentFind(vi []int, f2 func(int, int) bool) (index int, found bool) {
	boundOfVi := len(vi) - 1
	for i := 0; i < boundOfVi; i++ {
		if f2(vi[i], vi[i+1]) {
			return i, true
		}
	}
	return -1, false
}

func Search(scope []int, subScope []int, f2 func(int, int) bool) (value int,
	found bool) {
	lenOfScope := len(scope)
	lenOfSubScope := len(subScope)
	i := 0
	if i > lenOfScope-lenOfSubScope {
		return -1, false
	}
	for {
		for j := 0; ; j++ {
			if j == lenOfSubScope {
				return i, true
			}
			if !f2(scope[i+j], subScope[j]) {
				if i < lenOfScope-lenOfSubScope {
					i++
					break
				} else {
					return -1, false
				}
			}
		}
	}
}

func SearchN(scope []int, count int, match int, f2 func(int, int) bool) (
	value int, found bool) {
	lenOfScope := len(scope)
	i := 0
	if i > lenOfScope-count {
		return -1, false
	}
	for {
		for j := 0; ; j++ {
			if j == count {
				return i, true
			}
			if !f2(scope[i+j], match) {
				if i < lenOfScope-count {
					i++
					break
				} else {
					return -1, false
				}
			}
		}
	}
}

File: lua/non_modifying_sequence.lua
local function allOf(seq, func)
    for i = 1, #seq do
        if not func(seq[i]) then return false end
    end
    return true
end

local function anyOf(seq, func)
    for i = 1, #seq do
        if func(seq[i]) then return true end
    end
    return false
end

local function noneOf(seq, func)
    return not anyOf(seq, func)
end

local function forEach(table, func)
    for _, v in pairs(table) do
        func(v)
    end
end

local function countIf(table, func)
    local count = 0
    for _, v in pairs(table) do
        if func(v) then count = count + 1 end
    end
    return count
end

local function count(table, match)
    return countIf(table, function(v)
        return v == match
    end)
end

local function mismatch(s1, s2)
    local min = #s1 > #s2 and #s2 or #s1
    for i = 1, min do
        if s1[i] ~= s2[i] then return i end
    end
    return min
end

local function equal(t1, t2)
    if #t1 ~= #t2 then return false end
    for k, v in pairs(t1) do
        if t2[k] ~= v then return false end
    end
    return true
end

-- use nil to indicate not found
local function findIf(table, func)
    for k, v in pairs(table) do
        if func(k, v) then return k end
    end
    return nil
end

local function findIfNot(table, func)
    for k, v in pairs(table) do
        if not func(k, v) then return k end
    end
    return nil
end

local function find(table, match)
    return findIf(table, function(k, v)
        return v == match
    end)
end

-- use 0 to indicate not found
local function index(seq, match, start)
    for i = start, #seq do
        if seq[i] == match then return i end
    end
    return 0
end

local function indexReversed(seq, match, endAt)
    for i = endAt, 1, -1 do
        if seq[i] == match then return i end
    end
    return 0
end

local function findEnd(seq, sub)
    if #sub == 0 then return 0 end
    local mismatch = false
    local start = indexReversed(seq, sub[#sub], #seq) - #sub
    while start >= 0 do -- if not found, start == -#sub
        for i = 1, #sub do
            if seq[start+i] ~= sub[i] then
                mismatch = true
                break
            end
        end
        if not mismatch then return start + 1 end
        start = indexReversed(seq, sub[#sub], start + #sub - 1) - #sub
    end
    return 0
end

local function findFirstOf(seq, sub)
    for i = 1, #sub do
        local met = index(seq, sub[i], 1)
        if met ~= 0 then return met end
    end
    return 0
end

local function search(seq, sub)
    if #sub == 0 then return 0 end
    local mismatch = false
    local start = index(seq, sub[1], 1)
    while start > 0 do
        for i = 1, #sub do
            if seq[start+i-1] ~= sub[i] then
                mismatch = true
                break
            end
        end
        if not mismatch then return start end
        start = index(seq, sub[1], start + 1)
    end
    return 0
end


local function searchN(seq, n, value)
    local sub = {}
    for i = 1, n do
        table.insert(sub, value)
    end
    return search(seq, sub)
end

return {
    allOf = allOf,
    anyOf = anyOf,
    noneOf = noneOf,
    forEach = forEach,
    count = count,
    countIf = countIf,
    mismatch = mismatch,
    equal = equal,
    find = find,
    findIf = findIf,
    findIfNot = findIfNot,
    findEnd = findEnd,
    findFirstOf = findFirstOf,
    search = search,
    searchN = searchN,
}
